### PR TITLE
last_kept & last_viewed to libraries

### PR DESCRIPTION
**Pull request by ahsu1230**
@andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala [hunk 1]
```diff
@@ -103,11 +105,11 @@ class LibraryRepoImpl @Inject() (
     getByNameOrSlugCompiled(userId, name, slug, excludeState).firstOption
   }
 
-  def getByUser(userId: Id[User], excludeState: Option[State[Library]], excludeAccess: Option[LibraryAccess])(implicit session: RSession): Seq[(LibraryAccess, Library)] = {
+  def getByUser(userId: Id[User], excludeState: Option[State[Library]], excludeAccess: Option[LibraryAccess])(implicit session: RSession): Seq[(LibraryMembership, Library)] = {
     val q = for {
       lib <- rows if lib.state =!= excludeState.orNull
       lm <- libraryMembershipRepo.rows if lm.libraryId === lib.id && lm.userId === userId && lm.access =!= excludeAccess.orNull && lm.state === LibraryMembershipStates.ACTIVE
-    } yield (lm.access, lib)
+    } yield (lm, lib)
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala:113]: getByUser now returns LibraryMembership instead of just LibraryAccess so we can also have `lastViewed` field
Also will give us more flexibility in the future if we'd like to include other things from LibraryMembership

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala:113]: Sounds good.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala [hunk 2]
```diff
@@ -216,6 +218,7 @@ class KeepInterner @Inject() (
       // A inactive keep may have had tags already. Index them if any.
       keepToCollectionRepo.getCollectionsForKeep(internedKeep.id.get) foreach { cid => collectionRepo.collectionChanged(cid) }
     }
+    libraryRepo.save(library.copy(lastKept = Some(DateTime.now())))
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: Mind moving this to `internRawBookmarksWithStatus` and `internRawBookmark`? It would be nice to only have one public "gateway" method, but looks like there's two. Right now, the library will be updated many times if there's a batch operation.

Also, use an injected `Clock` instead of `DateTime.now()`, so that it's easier to test (in tests, we can set specific values that the injected clock will return, so we can have deterministic outcomes and specifically check edge cases).

**Comment by andrewconner** [comment]
Looks good. Minor comments, feel free to merge when ready.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala (hunk 2 above)
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: is it enough to just UPDATE the `lastKept` field, instead of SAVE the entire object, hence bump up seqNum? Services like search, cortex, curator are all listening on this seqNum. For example, `getLibrariesAndMembershipsChanged` in `ShoeboxDataPipeController`. When this seqNum changed, that method also query `libMemshipRepo`. It's expensive. If we can avoid unnecessary seqNum bump up, that would be great. 

@leogrim  @ymatsuda  @stkem

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 3]
```diff
@@ -107,24 +108,31 @@ class LibraryController @Inject() (
...
+          libraryRepo.getBySlugAndUserId(userId = owner.id.get, slug = LibrarySlug(slugStr))
+        } match {
+          case None =>
+            Future.successful(BadRequest(Json.obj("error" -> "no_library_found")))
+          case Some(lib) =>
+            if (libraryCommander.canViewLibrary(request.userIdOpt, lib, authToken, passPhrase)) {
+              request.userIdOpt.map { userId =>
+                db.readWrite { implicit s =>
+                  libraryMembershipRepo.getWithLibraryIdAndUserId(lib.id.get, userId).map { mem =>
+                    libraryMembershipRepo.save(mem.copy(lastViewed = Some(DateTime.now)))
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: again, is it necessary to save the entire object here? I am not feeling strongly against this.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: Agreed. There should be a special `updateLastKeptTo` method that specifically does not update the seqNum. Good catch.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: Make sure to add a comment saying that we're purposely not incrementing the seqNum, so someone later doesn't think it was accidentally left out.
